cardano-db-sync/src/Cardano/DbSync/OffChain/Query.hs

```diff
@@ -0,0 +1,223 @@
+  PoolMetadataRef,
+  PoolMetadataRefId,
+  PoolUrl,
+  VoteMetaHash (..),
+  VoteUrl,
+  VotingAnchor,
+  VotingAnchorId,
+ )
+import Cardano.DbSync.OffChain.FetchQueue (newRetry, retryAgain)
+import Cardano.DbSync.Types (OffChainPoolWorkQueue (..), OffChainVoteWorkQueue (..))
+import Cardano.Prelude hiding (from, groupBy, on, retry)
+import Data.Time (UTCTime)
+import Data.Time.Clock.POSIX (POSIXTime)
+import qualified Data.Time.Clock.POSIX as Time
+import Database.Esqueleto.Experimental (
+  SqlBackend,
+  SqlExpr,
+  Value (..),
+  ValueList,
+  desc,
+  from,
+  groupBy,
+  in_,
+  innerJoin,
+  just,
+  max_,
+  notExists,
+  on,
+  orderBy,
+  select,
+  subList_select,
+  table,
+  where_,
+  (:&) ((:&)),
+  (==.),
+  (^.),
+ )
+import System.Random.Shuffle (shuffleM)
+
+{- HLINT ignore "Fuse on/on" -}
+
+---------------------------------------------------------------------------------------------------------------------------------
+-- Query OffChain VoteData
+---------------------------------------------------------------------------------------------------------------------------------
+getOffChainVoteData :: MonadIO m => POSIXTime -> Int -> ReaderT SqlBackend m [OffChainVoteWorkQueue]
+getOffChainVoteData now maxCount = do
+  xs <- queryNewVoteWorkQueue now
+  if length xs >= maxCount
+    then take maxCount <$> liftIO (shuffleM xs)
+    else do
+      ys <- queryOffChainVoteWorkQueue (Time.posixSecondsToUTCTime now)
+      take maxCount . (xs ++) <$> liftIO (shuffleM ys)
+
+-- get all the voting anchors that don't already exist in OffChainVoteData or OffChainVoteFetchError
+queryNewVoteWorkQueue :: MonadIO m => POSIXTime -> ReaderT SqlBackend m [OffChainVoteWorkQueue]
+queryNewVoteWorkQueue now = do
+  res <- select $ do
+    va <- from $ table @VotingAnchor
+    where_ (notExists $ from (table @OffChainVoteData) >>= \ocvd -> where_ (ocvd ^. OffChainVoteDataVotingAnchorId ==. va ^. VotingAnchorId))
+    where_ (notExists $ from (table @OffChainVoteFetchError) >>= \ocvfe -> where_ (ocvfe ^. OffChainVoteFetchErrorVotingAnchorId ==. va ^. VotingAnchorId))
+    pure
+      ( va ^. VotingAnchorId
+      , va ^. VotingAnchorDataHash
+      , va ^. VotingAnchorUrl
+      )
+  pure $ map convert res
+  where
+    convert :: (Value VotingAnchorId, Value ByteString, Value VoteUrl) -> OffChainVoteWorkQueue
+    convert (Value vaId, Value vaHash, Value url) =
+      OffChainVoteWorkQueue
+        { oVoteWqMetaHash = VoteMetaHash vaHash
+        , oVoteWqReferenceId = vaId
+        , oVoteWqRetry = newRetry now
+        , oVoteWqUrl = url
+        }
+
+queryOffChainVoteWorkQueue :: MonadIO m => UTCTime -> ReaderT SqlBackend m [OffChainVoteWorkQueue]
+queryOffChainVoteWorkQueue _now = do
+  res <- select $ do
+    (va :& ocpfe) <-
+      from
+        $ table @VotingAnchor
+          `innerJoin` table @OffChainVoteFetchError
+        `on` (\(va :& ocpfe) -> ocpfe ^. OffChainVoteFetchErrorVotingAnchorId ==. va ^. VotingAnchorId)
```

Comment:
Similarly to `queryOffChainPoolWorkQueue`, we need to check here that there is no entry for this at `OffChainVoteData`.